Implementation/62363 adapt attachment uploads on wp comments so that they are attached to the journal container on submit

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

- [x] https://community.openproject.org/wp/62363
- [x] https://community.openproject.org/wp/62364

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Change the (wp) comment attachment from the work package to the journal

_[Long video]_

https://github.com/user-attachments/assets/16a1c367-cadc-4291-bc2f-0596952e5627

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

Change the CKEditor resource from the `WorkPackage` to the `Journal` such that:
* The journal is a new record when creating a new comment, any attachments uploaded then are uncontainered until the comment is submitted at which point they are linked to the now persisted journal
* In edit state, new attachments are assigned to the Journal
* The comment inline attachments are not displayed on the Files Tab. Their lifecycle is tied to the comment. _(NB: edit case lifecycle handled separately https://community.openproject.org/wp/63384)_

CKEditor is also updated to allow mentions on `Activity::Comment` https://github.com/opf/commonmark-ckeditor-build/pull/102

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
Listen to attachment added events for form submission

akabiru committed:
Make journals attachable in a basic form

akabiru committed:
Define journal attachable spec

akabiru committed:
Derive comment attachment claims from rich text content

Claim inline attachments that are actually in use- leave any previosly uploaded but unclaimed
attachemnts for clean up (later)- this way we don't clog up storage with unused image files

akabiru committed:
Remove client side attachment event handling

akabiru committed:
Resolve eslint issue

akabiru committed:
Add comment attachment claims contract spec

akabiru committed:
Add attachments claims set attribute service spec

akabiru committed:
Add activity tab image upload feature spec

akabiru committed:
Specify inline attachment in spec

akabiru committed:
Allow attachments to be viewable by other authorized users

akabiru committed:
Add corresponding methods for activity dynamic models
* `Activity` -> attachments_by_activity
* `Activity::Comment` -> attachments_by_activity_comment
* `Activity::BcfComment` -> attachments_by_activity_bcf_comment

akabiru committed:
Reuse existing attachment replacements override

Further adopt replacements context so that only attachments in use
are ever assigned to the journal

akabiru committed:
Use a single journal to avoid validity issues

akabiru committed:
Add restricted context to journal visibility on API

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

akabiru committed:
Merge branch 'implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit' of https://github.com/opf/openproject into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

akabiru committed:
Add unit tests for activities api with restricted journals cases

akabiru committed:
Add attachment by activity comment API spec

Support contaner aliasing needed from "journal" to "activity"

akabiru committed:
Add swagger documentation for activity attachment APIs

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

akabiru committed:
Embed workpackage in activity for mentions capability

See: https://github.com/opf/commonmark-ckeditor-build/pull/102

Co-authored-by: Bruno Pagno <b.pagno@openproject.com>

akabiru committed:
Ignore duplicate definition, separate interface and constant

akabiru committed:
Remove redundant specs

Attachment scenarios are covered in `spec/features/activities/work_package/activity_tab_comment_editor_spec.rb`
Notably: there is no longer a special permission for adding attachments to wp comments, rather they are treated
as inline- if a user can comment, then they can add attachments.

The lifecycle of the attachment is tied to it's containing comment.

akabiru commented:
Custom actions known flaky test: https://github.com/opf/openproject/actions/runs/14447556777/job/40511795904 https://github.com/opf/openproject/pull/18530

akabiru commented:
This pull request introduces several enhancements and new features related to handling attachments in the activities tab of work packages. The most important changes include the addition of new services for managing comment attachments, updates to the API schema to support attachment operations, and modifications to the frontend to handle attachment links.

Enhancements to attachment handling:

* [`app/contracts/work_packages/activities_tab/comment_attachments_claims_contract.rb`](diffhunk://#diff-d122123f3cfa391f97024ea89678d9c64ab25c45a693d6c239c721a87bc1842cR1-R33): Added a new contract class `CommentAttachmentsClaimsContract` to validate attachment replacements.
* [`app/services/work_packages/activities_tab/comment_attachments_claims/claims_service.rb`](diffhunk://#diff-5b04692b08e9a6a6dea4a1b81189c75ec9d4a605f0e720f51b14efffbacf9204R1-R45): Introduced `ClaimsService` to handle the claiming of journal attachments.
* [`app/services/work_packages/activities_tab/comment_attachments_claims/set_attributes_service.rb`](diffhunk://#diff-f5a8e0aedb9fb3ab733155f7676db0de74f522e2453efc8517ea210f6744fe70R1-R67): Added `SetAttributesService` to set attributes for comment attachments, including parsing attachment IDs from notes.

API schema updates:

* [`docs/api/apiv3/components/schemas/activity_model.yml`](diffhunk://#diff-7a336556c25c94462171e1dd6713b513da2925fcd83569c627730a6bccc76fb7R5-R21): Updated the activity model schema to include embedded attachments and work package properties, as well as links for attachment operations. [[1]](diffhunk://#diff-7a336556c25c94462171e1dd6713b513da2925fcd83569c627730a6bccc76fb7R5-R21) [[2]](diffhunk://#diff-7a336556c25c94462171e1dd6713b513da2925fcd83569c627730a6bccc76fb7R52-R183) [[3]](diffhunk://#diff-7a336556c25c94462171e1dd6713b513da2925fcd83569c627730a6bccc76fb7R194-R201)
* [`docs/api/apiv3/paths/activity_attachments.yml`](diffhunk://#diff-b526785755e3151cd2b3d001aa38def2996244fdb2ab1807e09fa049773d8f8dR1-R217): Added new API paths for listing and adding attachments to activities.

Frontend modifications:

* [`frontend/src/app/core/state/attachments/attachments.service.ts`](diffhunk://#diff-7980539328953d2b0f67ac0d8d98bf842098f8885d2ffcce1d807110d05a88feR147-R155): Added a static method `getAttachmentsSelfLink` to retrieve the self-link for attachments from a resource.

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

brunopagno commented on app/services/work_packages/activities_tab/comment_attachments_claims/set_attributes_service.rb:
```diff
@@ -0,0 +1,67 @@
+        def collect_attachment_ids_from_notes
+          return [] if model.notes.blank?
+
+          parser.css("img.op-uc-image").filter_map do |img|
```
optional: I'd like to make this selector more explicit.

```ruby
ATTACHMENT_CSS_SELECTOR = "img.op-uc-image"
```

I expect this to be covered by tests in case the selector ever changes, but it is also a bit of a magic string.

brunopagno reviewed (approved):
This works pretty much how I would expect it to work. I tested locally and uploads, copy paster, quotes are doing what's expected.

Really awesome job :clap:

I left one optional comment there just for explicitness, but there was nothing that called my attention as possibly an issue.

That said I am not especially knowledgeable about the API spec .yml files. Would be nice to get a second pair of eyes double check those.

akabiru committed:
Increase wait time for custom action title assertion

There seems to be a delay in when the title is updated to the custom action description

akabiru committed:
Merge pull request #18643 from opf/flaky-test/custom-actions-spec

Increase wait time for custom action title assertion

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

akabiru committed:
Define img selector more explicitly

Co-authored-by: Bruno Pagno <b.pagno@openproject.com>

akabiru committed:
Merge branch 'dev' into implementation/62363-adapt-attachment-uploads-on-comments-so-that-they-are-attached-to-the-container-on-submit

Kharonus committed:
[#62363] updated and formatted API spec

akabiru commented:
Thank you for the API spec refinements @Kharonus ! 🤩
